File: frontend/src/types/planner.ts
export interface PlanRequest {
  city: string;
  budget: number;
  available_time: string;
  mood: string;
  interests: string[];
  constraints: string[];
}

export interface TimelineItem {
  start: string;
  end: string;
  type: string;
  name: string;
  cost: number;
  why: string;
  area?: string;
  crowd_level?: string;
}

export interface ValidationStatus {
  budget_ok: boolean;
  time_ok: boolean;
  constraints_ok: boolean;
  details?: string[];
}

export interface FinalPlan {
  title: string;
  summary: string;
  total_cost: number;
  remaining_budget: number;
  total_duration_minutes: number;
  timeline: TimelineItem[];
  tradeoffs: string[];
  validation: ValidationStatus;
  is_fallback?: boolean;
  fallback_reason?: string;
}

export interface TraceEvent {
  node: string;
  status: 'running' | 'completed' | 'failed' | 'skipped';
  message: string;
  duration_ms: number;
  data?: Record<string, any>;
  timestamp?: number;
}

export interface PlanResponse {
  session_id: string;
  plan: FinalPlan | null;
  needs_clarification: boolean;
  clarification_question?: string;
  trace: TraceEvent[];
  is_mock_data: boolean;
}

File: backend/app/schemas/planner.py
from pydantic import BaseModel, Field
from typing import List, Optional, Dict, Any


class PlanRequest(BaseModel):
    city: str = Field(..., description="City name (e.g. Bangalore)")
    budget: float = Field(..., description="Total budget in INR (e.g. 2000)")
    available_time: str = Field(..., description="Available duration (e.g. '4 hours', '2 hours', 'custom')")
    mood: Optional[str] = Field(default="", description="Current mood or vibe")
    interests: List[str] = Field(default_factory=list, description="List of user interests")
    constraints: List[str] = Field(default_factory=list, description="List of constraints or preferences")


class ClarifyRequest(BaseModel):
    session_id: str
    clarification: str


class TimelineItem(BaseModel):
    start: str
    end: str
    type: str  # activity, food, music, walk, etc.
    name: str
    cost: float
    why: str
    area: Optional[str] = None
    crowd_level: Optional[str] = None


class ValidationStatus(BaseModel):
    budget_ok: bool = True
    time_ok: bool = True
    constraints_ok: bool = True
    details: Optional[List[str]] = None


class FinalPlan(BaseModel):
    title: str
    summary: str
    total_cost: float
    remaining_budget: float
    total_duration_minutes: int
    timeline: List[TimelineItem]
    tradeoffs: List[str] = Field(default_factory=list)
    validation: ValidationStatus = Field(default_factory=ValidationStatus)
    is_fallback: bool = False
    fallback_reason: Optional[str] = None


class TraceEvent(BaseModel):
    node: str
    status: str  # running, completed, failed, skipped
    message: str
    duration_ms: int = 0
    data: Optional[Dict[str, Any]] = None


class PlanResponse(BaseModel):
    session_id: str
    plan: Optional[FinalPlan] = None
    needs_clarification: bool = False
    clarification_question: Optional[str] = None
    trace: List[TraceEvent] = Field(default_factory=list)
    is_mock_data: bool = True
